### DISCUSSION TITLE
Improve poor book photos for OCR

### DISCUSSION THREAD

**Discussion by wumbel** [ImageMagick/ImageMagick#6537] [Help]
### ImageMagick version

6.9

### Operating system, version and so on

Ubuntu 22.04 LTS

### Description

Hello,
i have make photos of a big book and these are poor in the reference to a OCR. The problem is a shadow in the photos. The shadow has a too big contrast. So, when the OCR is working, the part with the shadow are too dark. When i increase the brightness, then the part without shadow is too bright.

Yet i search for a method to reduce the contrast or respectively make only the dark shadow part brighter.

There is some limited knowledge with Imagemagick on my side. I use it very rarely since about 10 years, both in Linux (command line) and PHP.

**Comment by fmw42** [comment]
Post an example image

**Comment by snibgo** [comment]
A useful technique for a scanned page is to make an image that represents the paper with shadow but no printing. Then divide the scanned page by the paper image (or a slightly lighter image). This lightens the page that was in shadow, so the background paper becomes white (or slightly off-white).

**Comment by wumbel** [comment]
@snibgo it's not so easy why you think. The shadows are not always on the same place and there are very many images.

Examples (images flipped, because image search bots use OCR):
![001](https://github.com/ImageMagick/ImageMagick/assets/141268557/82740f4b-23f5-492d-8a80-fae464f0fad0)

![002](https://github.com/ImageMagick/ImageMagick/assets/141268557/e3eb1915-dfcb-450b-a456-eed593c44d95)

![003](https://github.com/ImageMagick/ImageMagick/assets/141268557/d6cf4fb2-8ea0-4fc7-9163-b4069ec14c49)

**Comment by snibgo** [comment]
This works well for all your examples. Windows syntax.
```
magick ^
  %1 ^
  ( +clone ^
    -statistic Maximum 15x15 ^
    -blur 0x3 ^
  ) ^
  -compose DivideSrc -composite ^
  -flop ^
  x.png
```
`%1` is the input image. We make a copy, and set each pixel to the maximum value in a sliding 15x15 window. This makes a blocky result, so we blur it to smooth the blockiness. (The blurring is probably not needed for OCR.) The result is an approximation to the paper with shadow but no text.

We divide the input image by the paper image. We `-flop` to put the text the right way round.

Tesseract reads the images well, but not perfectly. Further processing with morphology could remove isolated black dots from the image.

To process all the images in a directory, you could use a shell loop.

**Comment by fmw42** [comment]
Here is one way

![text](https://github.com/ImageMagick/ImageMagick/assets/13458218/e16841a4-4fc6-47f2-8415-05266ac78e7c)

`convert text.jpg -colorspace gray \( +clone -blur 0x99 \) +swap -compose divide -composite result.jpg`

![result](https://github.com/ImageMagick/ImageMagick/assets/13458218/67fa44a9-bc0b-4b4c-84ad-45b0b7eea9c6)

**Comment by wumbel** [comment]
@snibgo , Linux, and thanks for your help, but the solution of fmw42 ...

@fmw42 , wow, the result looks very great!
I haven't time to test it timely, but this result give me a hope to retry the OCR in the next months.
